registry/handlers/layerupload.go

```diff
@@ -198,98 +167,100 @@ func (luh *layerUploadHandler) PutLayerUploadComplete(w http.ResponseWriter, r *
+			httpErr = NewHTTPError(v2.ErrorCodeDigestInvalid, err, http.StatusBadRequest)
 		default:
-			ctxu.GetLogger(luh).Errorf("unknown error completing upload: %#v", err)
-			w.WriteHeader(http.StatusInternalServerError)
-			luh.Errors.Push(v2.ErrorCodeUnknown, err)
+			ctxu.GetLogger(ctx).Errorf("unknown error completing upload: %#v", err)
+			httpErr = NewHTTPError(v2.ErrorCodeUnknown, err, http.StatusInternalServerError)
 		}
 
 		// Clean up the backend layer data if there was an error.
-		if err := luh.Upload.Cancel(); err != nil {
+		if err := upload.Cancel(); err != nil {
 			// If the cleanup fails, all we can do is observe and report.
-			ctxu.GetLogger(luh).Errorf("error canceling upload after error: %v", err)
+			ctxu.GetLogger(ctx).Errorf("error canceling upload after error: %v", err)
 		}
 
-		return
+		return httpErr
 	}
 
 	// Build our canonical layer url
-	layerURL, err := luh.urlBuilder.BuildBlobURL(luh.Repository.Name(), layer.Digest())
+	layerURL, err := ctx.urlBuilder.BuildBlobURL(ctx.Repository.Name(), layer.Digest())
 	if err != nil {
-		luh.Errors.Push(v2.ErrorCodeUnknown, err)
-		w.WriteHeader(http.StatusInternalServerError)
-		return
+		return NewHTTPError(v2.ErrorCodeUnknown, err, http.StatusInternalServerError)
 	}
 
 	w.Header().Set("Location", layerURL)
 	w.Header().Set("Content-Length", "0")
 	w.Header().Set("Docker-Content-Digest", layer.Digest().String())
 	w.WriteHeader(http.StatusCreated)
+	return nil
 }
 
 // CancelLayerUpload cancels an in-progress upload of a layer.
-func (luh *layerUploadHandler) CancelLayerUpload(w http.ResponseWriter, r *http.Request) {
-	if luh.Upload == nil {
-		w.WriteHeader(http.StatusNotFound)
-		luh.Errors.Push(v2.ErrorCodeBlobUploadUnknown)
-		return
+func CancelLayerUpload(ctx *Context, w http.ResponseWriter, r *http.Request) error {
+	uuid, upload, _, httpErr := layerUploadInfo(ctx, r)
+	if httpErr != nil {
+		return httpErr
+	}
+
+	if upload == nil {
+		return NewHTTPError(v2.ErrorCodeBlobUploadUnknown, nil, http.StatusNotFound)
 	}
 
-	w.Header().Set("Docker-Upload-UUID", luh.UUID)
-	if err := luh.Upload.Cancel(); err != nil {
-		ctxu.GetLogger(luh).Errorf("error encountered canceling upload: %v", err)
-		w.WriteHeader(http.StatusInternalServerError)
-		luh.Errors.PushErr(err)
+	w.Header().Set("Docker-Upload-UUID", uuid)
+	if err := upload.Cancel(); err != nil {
+		ctxu.GetLogger(ctx).Errorf("error encountered canceling upload: %v", err)
+		return NewHTTPError(v2.ErrorCodeUnknown, err, http.StatusInternalServerError)
 	}
 
 	w.WriteHeader(http.StatusNoContent)
+	return nil
 }
 
 // layerUploadResponse provides a standard request for uploading layers and
 // chunk responses. This sets the correct headers but the response status is
 // left to the caller.
-func (luh *layerUploadHandler) layerUploadResponse(w http.ResponseWriter, r *http.Request) error {
+func layerUploadResponse(ctx *Context, upload distribution.LayerUpload, state *layerUploadState, w http.ResponseWriter) error {
```

Comment:
Yeah, I noticed state was frequently used when the info could be pulled straight from the layer. Will simplify.